substrate/primitives/state-machine/src/overlayed_changes/mod.rs

```diff
@@ -528,22 +532,23 @@ impl<H: Hasher> OverlayedChanges<H> {
 
 	/// Get an iterator over all child changes as seen by the current transaction.
 	pub fn children(
-		&self,
-	) -> impl Iterator<Item = (impl Iterator<Item = (&StorageKey, &OverlayedValue)>, &ChildInfo)> {
-		self.children.values().map(|v| (v.0.changes(), &v.1))
+		&mut self,
+	) -> impl Iterator<Item = (impl Iterator<Item = (&StorageKey, &mut OverlayedValue)>, &ChildInfo)>
+	{
+		self.children.values_mut().map(|v| (v.0.changes(), &v.1))
 	}
 
 	/// Get an iterator over all top changes as been by the current transaction.
-	pub fn changes(&self) -> impl Iterator<Item = (&StorageKey, &OverlayedValue)> {
+	pub fn changes(&mut self) -> impl Iterator<Item = (&StorageKey, &mut OverlayedValue)> {
```

Comment:
Pretty sure that all of these should be split into two functions: `changes` and `changes` mut.  
Otherwise the `mut` bleeds everywhere.  

Maybe there is also a smarter way where a `&mut self` can be casted into a `OverlayedChangesRef` or something and that then has the non-mutable functions.

Reply:
The new design involve all past '&' access are now '&mut' accesses because a value storage may modify itself upon read (a list of append operation may be concatenated to a single value on read).
This is the same for simple get storage functions.
I don't think I can do differently here (inner mutability would make no sense as we already have a single access to the overlay).